compiler/resolution/preFold.cpp

```diff
@@ -1731,6 +1733,110 @@ static Symbol* findMatchingEnumSymbol(Immediate* imm, EnumType* typeEnum) {
 }
 
 
+static Expr* unrollHetTupleLoop(CallExpr* call, Expr* tupExpr, Type* iterType) {
+  // Find the getIterator's statement and insert a no-op after it
+  // in order to create a place to insert new statements.
+  //
+  Expr* parentStmt = call->getStmtExpr();
+  CallExpr* noop = new CallExpr(PRIM_NOOP);
+  parentStmt->insertAfter(noop);
+
+  // Look for the associated ForLoop statement, removing all
+  // statements leading up to it (since they're unnecessary once
+  // we unroll).  This is very sensitive to the current IR format
+  // such that if it changes, we'll likely bail out of this
+  // prefold and fall into the "Heterogeneous tuples don't support
+  // this style of loop yet" error message in ChapelTuple.chpl.
+  //
+  ForLoop* theloop = NULL;
+  Expr* nextStmt = noop->next;
+  if (DeferStmt* defer = toDeferStmt(nextStmt)) {
+    nextStmt = nextStmt->next;
+    defer->remove();
+    if (BlockStmt* block = toBlockStmt(nextStmt)) {
+      nextStmt = nextStmt->next;
+      block->remove();
+      if (ForLoop* loop = toForLoop(nextStmt)) {
+        theloop = loop;
+          }
+        }
+  }
+
+  // assume the IR isn't well-formed until we find the loop index var
+  //
+  bool wellformed = false;
+
+  if (theloop != NULL) {
+
+    // Mark the loop's index variable as being 'const ref'.
+    // Ultimately, this should likely be 'ref' in some cases, but
+    // for now I'm erring on the side of avoiding any
+    // modifications to tuples rather than allowing modifications
+    // to tuple elements that shouldn't be modified.
+    //
+    Expr* firstStmt = theloop->body.first();
+    if (DefExpr* defexp = toDefExpr(firstStmt)) {
+      if (VarSymbol* loopVar = toVarSymbol(defexp->sym)) {
+        wellformed = true;
+        loopVar->addFlag(FLAG_REF_VAR);
+        loopVar->addFlag(FLAG_CONST);
+      }
+    }
+  }
+
+  // if something wasn't as we expected, bail out; we won't unroll
+  // the loop and will generate an error when we try to resolve
+  // the iterator on the heterogeneous tuple.
```

Comment:
Wouldn't it be easier to debug if you generate an error here with `INT_FATAL` ?

Reply:
Do you mean USR_FATAL?  (The error generated is user facing—essentially "this loop form isn't supported for heterogeneous tuples at present").

I do think you have a point.  The reason I did it this way was that I wasn't convinced I would be able to catch all cases where the compiler didn't recognize the loop pattern here (e.g., zippered iterations seem to be translated into very different IR depending on what kind of loop index variable pattern they used), so felt I'd need to rely on the module error code as a catch-all.  And then, having done so, I was nervous about trying to maintain similar or identical error messages in two places in the code.  I could put a note to developers by the module-level errors to point them back to this routine as a breadcrumb.

Reply:
The breadcrumb would be fine with me for now, thanks

Reply:
OK, did this in https://github.com/chapel-lang/chapel/pull/15034/commits/928c9eb049c7faa9b89ca676c7ca9d47be3f6aae